syft/pkg/cataloger/swift/cataloger.go

```diff
@@ -1,12 +1,17 @@
 /*
-Package swift provides a concrete Cataloger implementation for Podfile.lock files.
+Package swift provides a concrete Cataloger implementation for Podfile.lock and Package.resolved files.
 */
 package swift
 
 import (
 	"github.com/anchore/syft/syft/pkg/cataloger/generic"
 )
 
+func NewSwiftPackageManagerCataloger() *generic.Cataloger {
+	return generic.NewCataloger("spm-cataloger").
+		WithParserByGlobs(parsePackageResolved, "**/Package.resolved", "**/.package.resolved")
```

Comment:
Are there any other globs that we should consider for a swift project that would contain dependencies managed by the package manager?

I noticed the command `swift package show-dependencies` and didn't know if it sourced this graph from other local files.

Example - I ran `swift packages show-dependencies` for https://github.com/mozilla-mobile/firefox-ios
```
.
├── swift<https://github.com/danger/swift.git@3.16.0>
│   ├── logger<https://github.com/shibapm/Logger@0.2.3>
│   ├── version<https://github.com/mxcl/Version@2.0.1>
│   └── octokit.swift<https://github.com/nerdishbynature/octokit.swift@0.12.0>
│       └── requestkit<https://github.com/nerdishbynature/RequestKit.git@3.2.1>
└── danger-swift-coverage<https://github.com/f-meloni/danger-swift-coverage@1.2.1>
    └── swift<https://github.com/danger/swift.git@3.16.0>
        ├── logger<https://github.com/shibapm/Logger@0.2.3>
        ├── version<https://github.com/mxcl/Version@2.0.1>
        └── octokit.swift<https://github.com/nerdishbynature/octokit.swift@0.12.0>
            └── requestkit<https://github.com/nerdishbynature/RequestKit.git@3.2.1>
```

But that seems like it contains a lot less than what we resolved in the SBOM:
```
➜  firefox-ios git:(main) ✗ syft dir:. --name firefox-ios | grep swift
Flag --name has been deprecated, use: source-name
 ✔ Indexed file system                                                                                       .
 ✔ Cataloged packages              [678 packages]
Commandant                                                                     0.17.0                swift
CwlCatchException                                                              1.2.0                 swift
CwlPreconditionTesting                                                         1.2.0                 swift
DangerXCodeSummary                                                             1.1.0                 swift
Komondor                                                                       1.0.4                 swift
Komondor                                                                       1.1.4                 swift
Logger                                                                         0.2.3                 swift
Nimble                                                                         7.3.1                 swift
Nimble                                                                         8.0.5                 swift
OctoKit                                                                        0.10.0                swift
OctoKit                                                                        0.12.0                swift
PackageConfig                                                                  0.13.0                swift
PackageConfig                                                                  1.1.3                 swift
Quick                                                                          1.3.2                 swift
Quick                                                                          2.2.0                 swift
RequestKit                                                                     2.4.0                 swift
RequestKit                                                                     3.1.0                 swift
RequestKit                                                                     3.2.0                 swift
RequestKit                                                                     3.2.1                 swift
Rocket                                                                         1.0.0                 swift
Rocket                                                                         1.2.1                 swift
SWXMLHash                                                                      5.0.1                 swift
SWXMLHash                                                                      6.0.0                 swift
ShellOut                                                                       2.3.0                 swift
SourceKitten                                                                   0.28.0                swift
SourceKitten                                                                   0.32.0                swift
SwiftFormat                                                                    0.43.5                swift
SwiftFormat                                                                    0.50.5                swift
SwiftLint                                                                      0.38.2                swift
SwiftLint                                                                      0.48.0                swift
SwiftShell                                                                     5.0.1                 swift
SwiftShell                                                                     5.1.0-beta.1          swift
SwiftSyntax                                                                    0.50600.1             swift
SwiftyTextTable                                                                0.9.0                 swift
TestSpy                                                                        0.3.1                 swift
Version                                                                        1.2.0                 swift
Version                                                                        2.0.1                 swift
Yams                                                                           2.0.0                 swift
Yams                                                                           4.0.6                 swift
a-star                                                                         3.0.0-beta-1          swift
danger-swift                                                                   3.0.0                 swift
danger-swift-coverage                                                          1.2.1                 swift
dip                                                                            7.1.1                 swift
fuzi                                                                                                 swift
gcdwebserver                                                                                         swift
glean-swift                                                                    53.0.0                swift
ios_sdk                                                                        4.33.4                swift
kif                                                                            3.8.7                 swift
kingfisher                                                                     7.6.2                 swift
logger                                                                         0.2.3                 swift
mappamundi                                                                                           swift
octokit.swift                                                                  0.12.0                swift
requestkit                                                                     3.2.1                 swift
rust-components-swift                                                          117.0.20230711050401  swift
sentry-cocoa                                                                   8.8.0                 swift
snapkit                                                                        5.6.0                 swift
swift                                                                          3.16.0                swift
swift-argument-parser                                                          1.0.3                 swift
swift-snapshot-testing                                                         1.10.0                swift
swiftybeaver                                                                   2.0.0                 swift
swiftyjson                                                                     5.0.1                 swift
telemetry-ios                                                                  2.0.0                 swift
version                                                                        2.0.1                 swift
```

Also I noticed that this list is much longer a second time around after running a few `swift package` commands.

Is the user expected to have resolved/downloaded certain dependency information before generating the SBOM for this cataloger to produce the optimal package output?

Reply:
The user is expected to have resolved packages (this can be done by building an App which depends on SPM packages) otherwise a `Package.resolved` file is not generated.

Without knowing what you did in those swift package commands, it could be that you resolved more package information, but at a different level. My guess is that a target specific dependency got resolved at the root-project level, this part touches a bit on the question you had about the globs, as far as they go, it's my understanding that `.package.resolved` file contains the root level dependencies, whilst the `Package.resolved` files contain the packages that are specific to targets. So could be that one of the commands you added changed the scope of where the dependencies were resolved, is my best guess. I'm by no means an expert of swift package manager.